### ISSUE TITLE
Binding ARRAY of jsonb in Postgres function results in SQLDialectNotSupportedException

### ISSUE THREAD

**Issue by ciaccia** [jOOQ/jOOQ#7501]
Consider the following table-like function that takes an array of jsonb[] as parameter and returns a table with a single text column:

```sql
CREATE OR REPLACE FUNCTION foo(p_objs jsonb[])
RETURNS TABLE("t" text) AS $$
    SELECT v::text AS "t"
    FROM unnest(p_objs) as u(v);
$$ LANGUAGE SQL;
```

In plain SQL, you can call it for example as following:

```sql
SELECT * FROM foo(ARRAY['true'::jsonb,'{"bar":"foo"}']::jsonb[]);
```

I define in `jooq-conf.xml` a custom binding as following:

```xml
<forcedType>
    <userType>com.google.gson.JsonElement[]</userType>
    <binding>org.jooq.PostgresJsonbArrayBinder</binding>
    <expression>foo\.p_objs</expression>
    <types>_jsonb</types>
</forcedType>
```

The generated Java class looks perfect:

```java
/**
 * Call <code>foo</code>.
 */
public static Result<FooRecord> FOO(Configuration configuration, JsonElement[] pObjs) {
    return DSL.using(configuration).selectFrom(org.jooq.tables.Foo.FOO.call(pObjs)).fetch();
}
```

Whenever I try to read data from the function, an exception is thrown:

```java
JsonElement a = new Gson().fromJson("true", JsonElement.class);
JsonElement b = new Gson().fromJson("{\"bar\":\"foo\"}", JsonElement.class);
Result<FooRecord> res = dslCtx.selectFrom(FOO(new JsonElement[] { a, b })).fetch();
```

> [main] WARN org.jooq.exception.SQLDialectNotSupportedException - Not supported by dialect : Type class com.google.gson.JsonElement is not supported in dialect null
> org.jooq.exception.SQLDialectNotSupportedException: Type class com.google.gson.JsonElement is not supported in dialect null
> 	at org.jooq.impl.DefaultDataType.getDataType(DefaultDataType.java:855)

I was not able to make it working, whatever I tried. Am I missing something or jOOQ does not support this use case yet? The only way I managed to use was by manually generating the SQL to be executed, as following:

```java
Result<FooRecord> res = dslCtx.select().from("FOO(" + DSL.inline(PostgresUtils.toPGArrayString(new JsonElement[] { a, b })) + ")").fetchInto(FOO);
```

**Comment by lukaseder** [comment]
Thanks for your report. This looks like a duplicate of #6794. Can you confirm? I'll close this as duplicate. If it's not the same issue, please reopen and I'll look at it more in detail

**Comment by ciaccia** [comment]
Hallo Lukas, I'm not 100% sure this problem is exactly the same as #6794. The issue is for sure similar, but according to my current understanding they differ in the following. #6794's Table is generate as following:

    public Foo call(OffsetDateTime pA) {
        return new Foo(DSL.name(getName()), null, new Field[] { 
              DSL.val(pA, org.jooq.impl.SQLDataType.TIMESTAMPWITHTIMEZONE)
        });
    }

In my case the generated file looks as following:

    public Foo call(JsonElement[] pObjs) {
        return new Foo(DSL.name(getName()), null, new Field[] { 
              DSL.val(pObjs, org.jooq.impl.DefaultDataType.getDefaultDataType("jsonb").getArrayDataType())
        });
    }

The latter DataType always returns Object[], so the converter (even if present) would not be called. In the former, the returned DataType is correct, and could potentially be called.

Here it looks like that in both cases the converter is not considered, but in my case the information is lost even before the binding is executed.

**Comment by lukaseder** [comment]
You're right, thanks for detailing. I can reproduce the problem in fact. Will look into this right away.

**Comment by lukaseder** [comment]
Nope, it's the same as #6794. But thanks to this check, I found that #6794 was implemented incompletely :)
